runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -719,3 +756,719 @@ TR::TreeLowering::lowerArrayStoreCHK(TR::Node *node, TR::TreeTop *tt)
+   TR::SymbolReference *elementSymRef = comp->getSymRefTab()->findOrCreateArrayShadowSymbolRef(TR::Address, anchoredArrayBaseAddressNode);
+   TR::Node *elementLoadNode = TR::Node::createWithSymRef(comp->il.opCodeForIndirectArrayLoad(TR::Address), 1, 1, elementAddress, elementSymRef);
+   elementLoadNode->copyByteCodeInfo(node);
+
+   TR::TreeTop *elementLoadTT = NULL;
+   if (comp->useCompressedPointers())
+      elementLoadTT = originalBlock->append(TR::TreeTop::create(comp, TR::Node::createCompressedRefsAnchor(elementLoadNode)));
+   else
+      elementLoadTT = originalBlock->append(TR::TreeTop::create(comp, TR::Node::create(node, TR::treetop, 1, elementLoadNode)));
+
+   if (enableTrace)
+      traceMsg(comp, "Created array element load treetop n%dn node n%dn\n", elementLoadTT->getNode()->getGlobalIndex(), elementLoadNode->getGlobalIndex());
+
+
+   ///////////////////////////////////////
+   // 7. Store the return value from the array element load to the same register or temp used by the anchored node
+   TR::Node *storeArrayElementNode = createStoreNodeForAnchoredNode(anchoredNode, elementLoadNode, comp, enableTrace, "array element load");
+
+   elementLoadTT->insertAfter(TR::TreeTop::create(comp, storeArrayElementNode));
+
+   if (enableTrace)
+      traceMsg(comp, "Store array element load node n%dn to n%dn %s\n", elementLoadNode->getGlobalIndex(), storeArrayElementNode->getGlobalIndex(), storeArrayElementNode->getOpCode().getName());
+
+
+   ///////////////////////////////////////
+   // 8. Split (3) at the array element load node
+   TR::Block *arrayElementLoadBlock = originalBlock->split(elementLoadTT, cfg);
+
+   arrayElementLoadBlock->setIsExtensionOfPreviousBlock(true);
+
+   if (enableTrace)
+      traceMsg(comp, "Isolated array element load treetop n%dn node n%dn in block_%d\n", elementLoadTT->getNode()->getGlobalIndex(), elementLoadNode->getGlobalIndex(), arrayElementLoadBlock->getNumber());
+
+   int32_t dataWidth = TR::Symbol::convertTypeToSize(TR::Address);
+   if (comp->useCompressedPointers())
+      dataWidth = TR::Compiler->om.sizeofReferenceField();
+
+   TR::Node *arraylengthNode = TR::Node::create(TR::arraylength, 1, anchoredArrayBaseAddressNode);
+   arraylengthNode->setArrayStride(dataWidth);
+
+   elementLoadTT->insertBefore(TR::TreeTop::create(comp, TR::Node::createWithSymRef(TR::NULLCHK, 1, 1, arraylengthNode, comp->getSymRefTab()->findOrCreateNullCheckSymbolRef(comp->getMethodSymbol()))));
+
+   elementLoadTT->insertBefore(TR::TreeTop::create(comp, TR::Node::createWithSymRef(TR::BNDCHK, 2, 2, arraylengthNode, anchoredElementIndexNode, comp->getSymRefTab()->findOrCreateArrayBoundsCheckSymbolRef(comp->getMethodSymbol()))));
+
+
+   ///////////////////////////////////////
+   // 9. Create ificmpne node that checks classFlags
+   TR::SymbolReference *vftSymRef = comp->getSymRefTab()->findOrCreateVftSymbolRef();
```

Comment:
Can these next few lines leading up not be commoned up with the PR 12126 ?

Reply:
Sorry now I see what you mean. I can put them into `TR_J9VMBase` where `testIsClassValueType` has been created. It can be used by` lower{Load|Store}ArrayElement` and `lowerArrayStoreCHK`. 

```c++
TR::Node *
TR_J9VMBase::ifArrayCompValueType(TR::Node *arrayBaseAddressNode, bool isIfEq)
   {
   TR::SymbolReference *vftSymRef = TR::comp()->getSymRefTab()->findOrCreateVftSymbolRef();
   TR::SymbolReference *arrayCompSymRef = TR::comp()->getSymRefTab()->findOrCreateArrayComponentTypeSymbolRef();

   TR::Node *vft = TR::Node::createWithSymRef(node, TR::aloadi, 1, arrayBaseAddressNode, vftSymRef);
   TR::Node *arrayCompClass = TR::Node::createWithSymRef(node, TR::aloadi, 1, vft, arrayCompSymRef);
   TR::Node *testIsValueTypeNode = testIsClassValueType(arrayCompClass);
   TR::Node *ifNode = TR::Node::createif(isIfEq ? TR::ificmpeq : TR::ificmpne, testIsValueTypeNode, TR::Node::iconst(node, 0));

   return ifNode; 
   }
```

Reply:
Addressed in fb3fa06